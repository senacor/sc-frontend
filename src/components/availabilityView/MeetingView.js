import React from 'react';
import PropTypes from 'prop-types';
import Typography from '@material-ui/core/Typography';
import { withStyles } from '@material-ui/core/styles/index';
import * as actions from '../../actions';
import { connect } from 'react-redux';
import { getSelectedDate, getMeeting } from '../../reducers/selector';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import moment from 'moment';
import { extendMoment } from 'moment-range';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';

import { translateContent } from '../translate/Translate';

const styles = theme => ({
  container: {
    display: 'block !important',
    '& > *': {
      display: 'block !important',
      marginTop: 1 * theme.spacing.unit,
      marginBottom: 1 * theme.spacing.unit
    }
  },
  attendeeList: {
    '& ListItemText span': {
      fontSize: '0.875rem'
    }
  }
});

class MeetingView extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      location: 'Test',
      duration: 60
    };
  }

  componentDidMount() {
    this.props.fetchMeeting(1);
  }

  handleChange = name => event => {
    this.setState({
      [name]: event.target.value
    });
  };

  handleClickOfMeetingButton = event => {
    event.preventDefault();
    this.addMeeting();
  };

  addMeeting() {
    let startDateTime = moment(this.props.getSelectedDateTime);
    let start = startDateTime.utc().format('YYYY-MM-DDTHH:mmZ');
    let end = startDateTime
      .add(this.state.duration, 'm')
      .utc()
      .format('YYYY-MM-DDTHH:mmZ');
    let meeting_details = {
      //TODO replace hardcoded values
      prId: 1,
      start: start,
      end: end,
      location: this.state.location,
      //TODO replace hardcoded values
      requiredAttendeeIds: [4],
      optionalAttendeeIds: []
    };
    console.log(start);
    console.log(end);
    this.props.addMeeting(meeting_details);
  }

  render() {
    const { classes, meeting } = this.props;
    return (
      <div style={{ padding: 10 + 'px' }}>
        {meeting == null ? (
          <form className={classes.container} noValidate autoComplete="off">
            <TextField
              id="location"
              label="Ort"
              className={classes.textField}
              value={this.state.location}
              onChange={this.handleChange('location')}
              margin="normal"
            />
            <TextField
              id="duration"
              label="Dauer des Termins (min)"
              className={classes.textField}
              value={this.state.duration}
              onChange={this.handleChange('duration')}
              margin="normal"
            />
            <Button
              id="createMeeting"
              className={classes.button}
              variant="raised"
              onClick={this.handleClickOfMeetingButton}
              type="submit"
            >
              Termin erstellen
            </Button>
          </form>
        ) : (
          <div className={classes.container}>
            <Typography variant="headline">Termindetails</Typography>
            <Typography id="startDateTime" variant="body1">
              Start:{' '}
              {moment(meeting.start)
                .local()
                .format('DD-MM-YYYY HH:mm')}
            </Typography>
            <Typography id="endDateTime" variant="body1">
              Ende:{' '}
              {moment(meeting.end)
                .local()
                .format('DD-MM-YYYY HH:mm')}
            </Typography>
            <Typography variant="body1">
              Ort:{' ' + meeting.location}
            </Typography>
            <Typography variant="body2" className={classes.title}>
              Ben√∂tigte Teilnehmer
            </Typography>
            <div className={classes.attendeeList}>
              <List id="requiredAttendees">
                {Array.from(meeting.requiredAttendees).map(requiredAttendee => {
                  return (
                    <ListItem key={requiredAttendee.email} dense>
                      <ListItemText
                        primary={
                          requiredAttendee.name +
                          ' <' +
                          requiredAttendee.email +
                          '>'
                        }
                        secondary={
                          'Status: ' + translateContent(requiredAttendee.status)
                        }
                      />
                    </ListItem>
                  );
                })}
              </List>
            </div>
            {Array.from(meeting.optionalAttendees).length > 0 ? (
              <div className={classes.attendeeList}>
                <Typography variant="body2" className={classes.title}>
                  Optionale Teilnehmer
                </Typography>
                <List id="optionalAttendees">
                  {Array.from(meeting.optionalAttendees).map(
                    optionalAttendee => {
                      return (
                        <ListItem key={optionalAttendee.email}>
                          <ListItemText
                            primary={
                              optionalAttendee.name +
                              ' <' +
                              optionalAttendee.email +
                              '>'
                            }
                            secondary={
                              'Status: ' +
                              translateContent(optionalAttendee.status)
                            }
                          />
                        </ListItem>
                      );
                    }
                  )}
                </List>
              </div>
            ) : (
              ''
            )}
          </div>
        )}
      </div>
    );
  }
}

MeetingView.propTypes = {
  classes: PropTypes.object.isRequired
};

export const StyledComponent = withStyles(styles)(MeetingView);
export default connect(
  state => ({
    meeting: getMeeting(state),
    getSelectedDateTime: getSelectedDate(state)
  }),
  {
    addMeeting: actions.addMeeting,
    fetchMeeting: actions.fetchMeeting
  }
)(StyledComponent);
