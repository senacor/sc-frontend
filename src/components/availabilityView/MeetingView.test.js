import React from 'react';
import { shallow } from 'enzyme';
import { StyledComponent } from './MeetingView';
import { createShallow } from '@material-ui/core/test-utils';

describe('MeetingView', () => {
  let shallow = createShallow({ dive: true });
  const mockFetchMeeting = jest.fn();

  const meeting = {
    start: '2018-08-06T09:51:00.000+0000',
    end: '2018-08-06T10:51:00.000+0000',
    location: 'Test',
    requiredAttendees: [
      {
        name: 'PR Mitarbeiter2',
        email: 'test.pr.mitarbeiter2@senacor.com',
        status: 'DECLINE'
      },
      {
        name: 'PR Mitarbeiter3',
        email: 'test.pr.mitarbeiter3@senacor.com',
        status: 'ACCEPT'
      }
    ],
    optionalAttendees: [],
    _links: {
      self: {
        href: 'http://localhost:8010/api/v1/meetings?prId=1'
      }
    }
  };

  it('should match snapshot', () => {
    let wrapper = shallow(
      <StyledComponent
        selectedDateTime={'2018-0s8-07T11:27+02:00'}
        fetchMeeting={mockFetchMeeting}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should display the create appointment form when meeting not exists', () => {
    let shallow = createShallow({ dive: true });
    let wrapper = shallow(
      <StyledComponent
        meeting={null}
        selectedDateTime={'2018-08-07T11:27+02:00'}
        fetchMeeting={mockFetchMeeting}
      />
    );

    expect(wrapper.find('#createMeeting')).toHaveLength(1);
  });

  it('should display meeting details when a meeting exists', () => {
    let wrapper = shallow(
      <StyledComponent
        meeting={meeting}
        selectedDateTime={'2018-08-07T11:27+02:00'}
        fetchMeeting={mockFetchMeeting}
      />
    );

    expect(wrapper.find('#requiredAttendees').children()).toHaveLength(
      meeting.requiredAttendees.length
    );

    expect(wrapper.find('#optionalAttendees').children()).toHaveLength(
      meeting.optionalAttendees.length
    );
  });
});
