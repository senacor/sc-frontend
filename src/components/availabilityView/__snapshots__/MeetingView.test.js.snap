// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeetingView should match snapshot 1`] = `
<div
  style={
    Object {
      "padding": "10px",
    }
  }
>
  <form
    autoComplete="off"
    className="MeetingView-container-1"
    noValidate={true}
  >
    <TextField
      id="location"
      label="Ort"
      margin="normal"
      onChange={[Function]}
      required={false}
      select={false}
      value="Test"
    />
    <TextField
      id="duration"
      label="Dauer des Termins (min)"
      margin="normal"
      onChange={[Function]}
      required={false}
      select={false}
      value={60}
    />
    <WithStyles(Button)
      id="createMeeting"
      onClick={[Function]}
      type="submit"
      variant="raised"
    >
      Termin erstellen
    </WithStyles(Button)>
  </form>
</div>
`;
